Add edit username 1197779

jwhitlock opened the pull request:
Add username back to UserEditForm as a required (non-blankable) field.  Replaces #3474

jezdez commented:
The reason we have the username set to `required=False` is because we have legacy usernames in the current database that don't pass the actual validation we've applied over time. This isn't ideal and a result of years of bad profile and code management, just wanted to mention it. I'm okay with making it a required field and effectively forcing users to chose a username that applies. To be clear, once they edit their user profile even only to change another field, the saving will fail because their existing username doesn't work with our regex.

jwhitlock commented:
@jezdez - my reading of the code is that legacy users get a different regex, but that pattern doesn't allow blank usernames either:

```
(env)vagrant@developer-local:~/src$ ./manage.py shell
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
Type "copyright", "credits" or "license" for more information.

IPython 1.2.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from kuma.users.constants import *

In [2]: USERNAME_LEGACY_REGEX.match('')

In [3]: USERNAME_LEGACY_REGEX.match('f')
Out[3]: <_sre.SRE_Match at 0x7fd76423a100>
```

groovecoder commented:
@jwhitlock :+1: there aren't any legacy users in the database with blank usernames. (There **are** some legacy users with blank email addresses.)

groovecoder commented on kuma/users/tests/test_views.py:
```diff
@@ -127,12 +127,13 @@ def tearDown(self):
-                                fn).val()
-        form['country'] = 'us'
-        form['format'] = 'html'
+            form[prefix + fn] = doc.find('#user-edit *[name="user-%s"]' %
```
nit: while we're in here, can we change `fn` to `fieldname` or `field_name` or something? 2-character variable names are only 50% less offensive than 1-character variable names. :smile:

jezdez commented:
@jwhitlock hm, *shrug*

jwhitlock commented:
I've rebased and made these changes:
* kuma/users/tests/test_views.py - Rewrote for style and to remove two-character variable name
* Added tests for legacy usernames
* Forbid blank usernames for everyone, and change validation error message

Coin flip says... @jezdez next

groovecoder commented on kuma/users/tests/test_views.py:
```diff
@@ -127,12 +127,15 @@ def tearDown(self):
+            'username', 'email', 'fullname', 'title', 'organization',
+            'location', 'irc_nickname', 'bio', 'interests')
+        for field in fields:
+            value = doc.find('#user-edit *[name="user-%s"]' % field).val()
```
nit: `value = doc.find('#user-edit *[name="%s-%s"]' % (prefix, field)).val()`?

groovecoder replied:
Though we could go on forever with using the prefix var instead of hard-coding. Meh. Up to you if it bothers you enough ...

jwhitlock commented:
OK, now it bothers me.

jwhitlock commented:
* Re-wrote ``_get_current_form_field_values`` again

Assigning to @groovecoder so that @jezdez will comment

groovecoder commented:
flake8 failure in Travis:
```
kuma/users/tests/test_forms.py:9:1: F811 redefinition of unused 'USERNAME_CHARACTERS' from line 8
```

groovecoder commented on kuma/users/tests/test_views.py:
```diff
@@ -126,13 +126,16 @@ def tearDown(self):
-                                fn).val()
-        form['country'] = 'us'
-        form['format'] = 'html'
+        lookup_pattern = '#{prefix}edit *[name="{prefix}{field}"]'
```
nice :+1:

jezdez replied:
MODERN PYTHON HOW DARE YOU USE SOMETHING NEW!!1!

jezdez replied:
^W^W^W^W :pray:

groovecoder commented:
r+ with green Travis

jezdez commented:
Looking good. r+

darkwing commented:
Kicking tests

jwhitlock commented:
Rebased again, and removed second import.  If it goes green, merging.
